### PR TITLE
Create `Useful.get_in_default/3` issue #31

### PR DESCRIPTION
**Pull request by nelsonic**
Fixes: 
+ [x] #31 
+ [x] #32 

Package published to https://hex.pm/packages/useful/1.10.0 📦 🚀

### CODE REVIEW

**Comment by nelsonic** [comment]
@LuchoTurtle this should be a quick one. 🤞 
Please review + merge when you can. [not urgent] 🙏 
Thanks. 🙌

**Comment by nelsonic** [comment]
@LuchoTurtle you can already review + merge this. 
But while I'm attempting to use it, I'm encountering https://github.com/dwyl/useful/issues/31#issuecomment-1360944448 ...
So I might be doing a bit more work ... ⏳

**Comment by LuchoTurtle** [comment]
I tried to get it working with [`Access.key/1`](https://hexdocs.pm/elixir/1.14.2/Access.html#key!/1) several times but I still got the same `Access behaviour not implemented error`. 
I tried converting the `conn` object in the `phoenix-todo-tutorial` project from a `struct` to a `map` and it did work. 

But to get it to work on all scenarios, I had to borrow a gist to convert [deeply nested structs to maps](https://gist.github.com/mpugach/2d264c063db9bc91af0f7c7ec8175037) (this by itself could actually be a function in this package, with appropriate credits) to consistently get this function to work.

I don't know if this is decent or if you don't want to borrow a small module to convert. I'll assign it back to you for review.

File: lib/useful.ex [hunk 1]
```diff
@@ -86,6 +86,35 @@ defmodule Useful do
...
+      180
+      iex> Useful.get_in_default(nil, [:unhappy, :path], "Happy!")
+      "Happy!"
+  """
+  def get_in_default(map, keys, default \\ nil) do
+    # Checking if it is a struct.
+    # If it is, we deeply convert it to map
+    map =
+      case Kernel.is_struct(map) do
+        true -> MapFromDeepStruct.from_deep_struct(map)
```
Comment (LuchoTurtle) [inline lib/useful.ex:108]: Check https://github.com/dwyl/useful/pull/33#issuecomment-1361268904 to verify if `MapFroMDeepStruct` is applicable.

**Review by LuchoTurtle — CHANGES_REQUESTED** [review]

**Comment by nelsonic** [comment]
@LuchoTurtle good detective work. The issue is testing the function. And given how frequently we would use this in our App, the constant conversion from Struct to Map would get “expensive”. 💭

**Comment by LuchoTurtle** [comment]
From what I've seen, to get anything similar to this to work would always involve a conversion, either with `Jason` or `Poison`. 

Using `Kernel.key`, as we already discussed in the standup, would involve iterating over the key array and reducing. But I couldn't get this to work.

Do you want me to scrap what I've done?

**Comment by nelsonic** [comment]
@LuchoTurtle you've added code to this PR and brought down the coverage:

```sh
Generated useful app
....................................................
Finished in 0.09 seconds (0.09s async, 0.00s sync)
21 doctests, 31 tests, 0 failures

Randomized with seed 587319
----------------
COV    FILE                                        LINES RELEVANT   MISSED
100.0% lib/useful.ex                                 261       38        0
 85.7% lib/utils.ex                                   23        7        1
[TOTAL]  97.8%
----------------
Generating report...
Saved to: cover/
```

image

# 💭

**Comment by LuchoTurtle** [comment]
Reverted the changes since the changes were too expensive to your liking.
I don't know why this is assigned to me since you said you "were going to be doing more work" 🤔

**Comment by LuchoTurtle** [comment]
Reverted the changes since the changes were too expensive to your liking.
I don't know why this is assigned to me since you said you "were going to be doing more work" 🤔

**Review by nelsonic — COMMENTED** [review]
Thanks @LuchoTurtle 👍

**Pull request merged**